### DISCUSSION TITLE
current state of documentation

### DISCUSSION THREAD

**Discussion by black-backdoor** [danielmiessler/Fabric#503] [Ideas]
Hey everyone,

I wanted to start a discussion about the current state of documentation in our project.

It's become apparent that there are some gaps and areas that need improvement, particularly regarding the setup process. One of the issues I've noticed is that all documentation seems to be centralized in the README.md file. There also seems to be a lack of documentation regarding the setup of the API Keys (issue #499). But I think we should keep the documentation simple.

What do you all think? I’d love to hear your thoughts.

**Comment by dhivah** [comment]
I would also recommend adding Readme file to every patterns with usage guide in it.

**Comment by YorkyPoo** [comment]
I'll start by saying that whenever I install a program manually, I tend to find a bug. Just about every AI program I installed found issues, likely due to my mistakes, but there were several instances where a developer had to make a hotfix for me. Even the one-click install for Pinokio didn't install everything properly and still gives me issues today.

I've just started running my own LLM's this year, but I've spent most of my life messing with computers, installing map hacks, and side-loading programs. Fabric was the first LLM I've installed, followed by Ollama. I had issues initially because there wasn't any documentation for installing on Windows. I didn't know that certain files needed to be made global, nothing mentioned installing Node.js/npm, and I could only get a compatible Python from the Windows store. This time, getting the GUI up and running, it took me several attempts. Ultimately, I had to start from scratch and provide all the instructions to Chat-GPT.

Anyway, here is the guide I had to follow:

"Sure! Here’s a comprehensive guide covering the entire process from installing Visual Studio Code to setting up and running the Fabric LLM framework with its GUI."

### 1. Install Visual Studio Code

1. **Download Visual Studio Code**:
    - Go to the [Visual Studio Code download page](https://code.visualstudio.com/).
    - Download and install the version appropriate for your operating system (Windows, macOS, Linux).

### 2. Install and Set Up WSL (Windows Users)

1. **Enable WSL**:
    - Open PowerShell as Administrator and run:
      ```sh
      wsl --install
      ```

2. **Install a Linux Distribution**:
    - After WSL is installed, you will be prompted to install a Linux distribution (e.g., Ubuntu) from the Microsoft Store. Follow the prompts to complete the installation.

3. **Set Up Ubuntu in WSL**:
    - Open the installed Ubuntu app.
    - Create a new user account when prompted.

### 3. Open VSCode and Set Up the Environment

1. **Open VSCode**.
2. **Open a New WSL Terminal**:
    - Go to the menu and select `Terminal` > `New Terminal`.
    - In the terminal, select your WSL distribution (e.g., Ubuntu).

### 4. Clone the Fabric Repository

1. **Navigate to Your Code Directory**:
    - In the WSL terminal:
      ```sh
      cd /mnt/d/Code/Fabric  # Change the path to your desired directory
      ```

2. **Clone the Fabric Repository**:
    ```sh
    git clone https://github.com/danielmiessler/fabric.git
    ```

3. **Enter the Fabric Directory**:
    ```sh
    cd fabric
    ```

### 5. Install pipx

1. **Update Your Package List**:
    ```sh
    sudo apt update
    ```

2. **Install pipx**:
    ```sh
    sudo apt install pipx
    ```

### 6. Install Fabric Using pipx

1. **Install Fabric**:
    ```sh
    pipx install .
    ```

2. **Run Setup**:
    ```sh
    fabric --setup
    ```

3. **Restart Your Shell**:
    ```sh
    exec $SHELL
    ```

### 7. Verify the Installation

1. **Check Fabric Installation**:
    ```sh
    fabric --help
    ```

### 8. Install Node.js and npm

1. **Install Node.js and npm**:
    ```sh
    sudo apt install nodejs npm
    ```

2. **Verify the Installation**:
    ```sh
    node -v
    npm -v
    ```

### 9. Set Up and Run the Fabric GUI

1. **Navigate to the Fabric Directory**:
    ```sh
    cd /mnt/d/Code/Fabric/fabric
    ```

2. **Install Fabric GUI Dependencies**:
    ```sh
    npm install
    ```

3. **Start the Fabric GUI**:
    ```sh
    fabric --gui
    ```

### Summary

Here’s the full command sequence for reference:

```sh
# Update your package list
sudo apt update

# Install Node.js and npm
sudo apt install nodejs npm

# Verify the installation
node -v
npm -v

# Navigate to the Fabric directory
cd /mnt/d/Code/Fabric/fabric

# Clone the Fabric repository
git clone https://github.com/danielmiessler/fabric.git

# Enter the Fabric directory
cd fabric

# Install pipx
sudo apt install pipx

# Install Fabric using pipx
pipx install .

# Run Fabric setup
fabric --setup

# Restart your shell
exec $SHELL

# Verify the installation
fabric --help

# Install Fabric GUI dependencies
npm install

# Start the Fabric GUI
fabric --gui
```

This guide should help you get the Fabric LLM framework up and running with its GUI. If you have any more questions or need further assistance, feel free to ask!

**Reply by Neurocoding** [reply]
did you get the drag and drop function in the gui to work ?

**Reply by YorkyPoo** [reply]
I did not know that was a function.

**Reply by fractaloutlook** [reply]
I followed the instructions off the readme and then just now through your list here.... and I can't get the gui to run...

npm install gives me:

![2024-06-09 17_23_37-Visual Studio Code  Administrator](https://github.com/danielmiessler/fabric/assets/27757547/75608923-b50e-4a55-be79-e2d6568bf921)

can't find the package.json? 

I'm fairly new to node.js but I've got all of that installed OK!

Trying to run the gui yields:

![2024-06-09 17_37_38-Visual Studio Code  Administrator](https://github.com/danielmiessler/fabric/assets/27757547/542c4543-4cb0-4e53-a7e2-170265660a5c)


Any thoughts would be appreciated! Don't really need the gui I suppoes I just wanted to see how it was!

**Comment by timothyjoh** [comment]
I think this project is considered "hacker friendly" at the moment, with the ability to make it "user friendly" probably coming at a later date, if at all.

Appreciate the Windows guide @YorkyPoo . I Intend to work on a Docker image way of getting it up and running, both locally as well as hosted on a server if anyone wants to do that as well.

**Reply by jakub-deka** [reply]
Docker image would be great. I have been trying to get this running under docker for a little while and I cannot get the api/web gui to work at all. I would appreciate some assistance if you guys have any ideas.

**Comment by black-backdoor** [comment]
## Proposal for Documentation Structure

After some thought, I came up with this file structure to split the documentation in the README into multiple files. Do you have any feedback?


/
├── README.md
├── docs/
│       ├── index.md
│       ├── getting-started.md
│       ├── helper-apps.md
│       ├── custom-patterns.md


#### getting-started.md
- setup API keys
- using Ollama server
- using Ollama locally
- using | & helper-apps
- setting up pbpaste on Linux

#### custom-patterns.md
- creating pattern folder
- copying patterns
- getting started writing your own patterns

#### README.md
- installation



Looking forward to your thoughts!

**Reply by YorkyPoo** [reply]
I think that would be useful. Anything to get more people set up easier and use fabric in more applications. Especially for people like me I just don't understand the framework behind everything. Even just recently I tried using the GUI. It was easy to use with smaller stuff but when trying to get it to do larger task it kept giving me errors. I know it's sometimes difficult to find the capabilities of the Ollama LLM. I was using GPT-3.5 because I know It can handle a lot of text.

**Reply by black-backdoor** [reply]
PR #554
